### PR TITLE
[dynamo] Introduce `get_real_value` API to TensorVariable

### LINKED ISSUE #1466: TORCHDYNAMO_DYNAMIC_SHAPES=1 expects sizes returned from graph to be tuples, but AOTAutograd doesn't agree
Repro: `TORCH_SHOW_CPP_STACKTRACES=1 TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 TORCH_MONKEYPATCH_GETITEM=1 AOT_FX_GRAPHS=1 python benchmarks/torchbench.py --only hf_T5 --accuracy --backend aot_eager --training)` with symbolic-shapes branch https://github.com/pytorch/pytorch/pull/84246

We end up with this AOTAutograd graph:

```
====== Forward graph ======                                                                                                                                
class GraphModule(torch.nn.Module):                                                                                                                        
    def forward(self, primals_1, primals_2, primals_3):                                                                                                    
                                                                                                                                                           
        # Module stack: {}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py:925, code: input_ids
 = input_ids.view(-1, input_shape[-1])                                                                                                                     
        sym_size = torch.ops.aten.sym_size(primals_2, 1)                                                                                                   
        view = torch.ops.aten.view.default(primals_2, [-1, sym_size])                                                                                      
                                                                             
        # Module stack: {'self_embed_tokens': 'Embedding'}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/mod
eling_t5.py:934, code: inputs_embeds = self.embed_tokens(input_ids)          
        embedding = torch.ops.aten.embedding.default(primals_1, view);  primals_1 = None                                                                   
                                                                             
        # Module stack: {}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py:945, code: attention
_mask = torch.ones(batch_size, mask_seq_length).to(inputs_embeds.device)                                                                                   
        sym_size_1 = torch.ops.aten.sym_size(primals_2, 0);  primals_2 = None                                                                              
        ones = torch.ops.aten.ones.default([sym_size_1, sym_size], device = device(type='cpu'), pin_memory = False)                              
        detach = torch.ops.aten.detach.default(ones);  ones = None                                                                                         
        detach_1 = torch.ops.aten.detach.default(detach);  detach = None                                                                                   
        _to_copy = torch.ops.aten._to_copy.default(detach_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  detach
_1 = None                                                                    
                                                                                                                                                           
        # Module stack: {}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py:948, code: encoder_a
ttention_mask = torch.ones(                                                                                                                                
        sym_size_2 = torch.ops.aten.sym_size(primals_3, 1);  primals_3 = None                                                                              
        ones_1 = torch.ops.aten.ones.default([sym_size_1, sym_size_2], dtype = torch.int64, device = device(type='cuda', index=0), pin_memory = False);  sy
m_size_2 = None                                                              
        detach_2 = torch.ops.aten.detach.default(ones_1);  ones_1 = None                                                                                   
        detach_3 = torch.ops.aten.detach.default(detach_2);  detach_2 = None                                                                               
        return [_to_copy, sym_size_1, sym_size, detach_3, embedding, view] 

aot output: (tensor([[1., 1., 1.,  ..., 1., 1., 1.],                                                                                                       
        [1., 1., 1.,  ..., 1., 1., 1.]], device='cuda:0',                                                                                                  
       grad_fn=<CompiledFunctionBackward>), 2, 1024, tensor([[1, 1, 1,  ..., 1, 1, 1],                                                                     
        [1, 1, 1,  ..., 1, 1, 1]], device='cuda:0'))  
```

but Dynamo is expecting (2, 1024)

Comments from @jansel :

Actually, perhaps we should just change `DynamicShapeVariable(len=2)` to `TupleVariable([DynamicShapeVariable, DynamicShapeVariable])`
There is actually a SizeVariable() in [lists.py](https://lists.py/). This is only a dynamic_shapes=True thing.
(dynamic_shapes=True is less tested)

We live debugged this on the stream at https://www.twitch.tv/edwardzyang

### PR DESCRIPTION
**Pull request by suo**
Right now, example_value is doing two jobs:
- We use it to propagate metadata (e.g. return type, shapes, etc.)
  throughout the graph
- We use it to satisfy queries for the actual value (e.g. torch.cond, `assume_constant_result`)

This is further complicated by the fact that we have two modes, one
where `example_value` is a fake tensor, and one where it is a real
tensor (this is the `fake_tensor_propagation` config flag).

This leads to scenarios where we don't support every combination of job + mode, 
e.g. if `fake_tensor_propagation=False`, `assume_constant_result` is broken.

This is made worse by the fact that "fake tensor mode" is the default
and is required if you want dynamic shapes to work.

So, this PR introduces a `get_real_value` API that just runs the graph
up to `node` in order to get a concrete value. This API is orthogonal to
`example_value`, so it doesn't care about `fake_tensor_propagation`.

When `fake_tensor_propagation=True`: `example_value` is a fake tensor,
you must use the `get_real_value` API to get a concrete value. This will
be the only configuration in the future.

When `fake_tensor_propagation=False`: `example_value` and
`get_real_value` will produce the same value. This is redundant but we
will be removing this config soon.

To support this, I introduce a cache for computed real values, to
memoize the work involved if we're asking for real values a lot.

I attached this state to `OutputGraph` because it seems to be what
historically managed `example_value` lifetimes, but idk.

### CODE REVIEW

**Comment by suo** [comment]
@voznesenskym I wanted to do the thing where I got rid of stashing `example_value` on the `node.meta`, but it turned out to be not clean in a variety of ways. So I'll defer it to a future PR.

**Review by jansel — APPROVED** [review]

**Review by voznesenskym — APPROVED** [review]

**Comment by jansel** [comment]
We have migrated `torchdynamo` to `torch._dynamo` and will use the `pytorch/pytorch` repo for future development.  Please resubmit this PR to https://github.com/pytorch/pytorch/

More details and instructions to port this PR over can be found in https://github.com/pytorch/torchdynamo/issues/1588
